packages/mongodb-dialects/java-driver/src/main/kotlin/com/mongodb/jbplugin/dialects/javadriver/glossary/JavaDriverDialectParser.kt

```diff
@@ -110,6 +90,40 @@ object JavaDriverDialectParser : DialectParser<PsiElement> {
             emptyList()
         }
 
+    override fun isReferenceToDatabase(source: PsiElement): Boolean {
+        val refToDb =
+            source
+                .parentOfType<PsiMethodCallExpression>(true)
+                ?.findMongoDbClassReference(source.project)
+                ?: return false
+
+        return refToDb.type?.isMongoDbDatabaseClass(refToDb.project) == true
+    }
+
+    override fun isReferenceToCollection(source: PsiElement): Boolean {
+        val refToDb =
+            source
+                .parentOfType<PsiMethodCallExpression>(true)
+                ?.findMongoDbClassReference(source.project)
+                ?: return false
+
+        return refToDb.type?.isMongoDbCollectionClass(refToDb.project) == true
+    }
+
+    override fun isReferenceToField(source: PsiElement): Boolean {
+        val isInQuery = isInQuery(source)
+        val isString = source.parentOfType<PsiLiteralExpression>()?.tryToResolveAsConstantString() != null
+
+        return isInQuery && isString
+    }
+
+    private fun isInQuery(element: PsiElement): Boolean {
+        val methodCall = element.parentOfType<PsiMethodCallExpression>(false) ?: return false
+        val containingClass = methodCall.resolveMethod()?.containingClass ?: return false
+
+        return containingClass.qualifiedName == FILTERS_FQN || containingClass.qualifiedName == UPDATES_FQN
```

Comment:
This entire logic looks so succinct compared to the one we had [before](https://github.com/mongodb-js/intellij/pull/40/files#diff-912f16e78c7f26e58229cd1de9b448c336085fbd923fe2fad6315e35d998db85L16-L36). Just confirming that we are not missing anything critical here.

Reply:
Yeah, one of the issues with this refactor that I had to do to align the behaviour with Spring Criteria is that I broke almost all the tests the first time, and had to play a lot around it to make sure the tests keep passing. Was a hard one.

I think it should not be breaking anything, and also did a lot of manual testing on top of our automated tests.

Reply:
Perfect 🙌